In [97]:
#Importing Libraries

import pandas as pd
import json
from datetime import datetime
import numpy as np
import sqlalchemy
import psycopg2

## First Dataset (receipts)

In [98]:
#Reading first dataset (receipts)

receipts = pd.read_json('./receipts.json', lines = True)

In [99]:
receipts

_id  bonusPointsEarned  \
0     {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1     {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2     {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3     {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4     {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   
5     {'$oid': '5ff1e1e40a7214ada1000566'}              750.0   
6     {'$oid': '5ff1e1cd0a720f052300056f'}                5.0   
7     {'$oid': '5ff1e1a40a720f0523000569'}              500.0   
8     {'$oid': '5ff1e1ed0a7214ada100056e'}                5.0   
9     {'$oid': '5ff1e1eb0a7214ada100056b'}              250.0   
10    {'$oid': '5ff1e1c50a720f052300056c'}              100.0   
11    {'$oid': '5ff1e1a10a720f0523000568'}              750.0   
12    {'$oid': '5ff1e1b60a7214ada100055c'}              150.0   
13    {'$oid': '5f9c74f70a7214ad07000037'}              750.0   
14    {'$oid': '5ff1e1b20a7214ada100055a'}              300.0   
15    {'$oid': '5ff1e1e90a7214ada1000569'}                NaN   
16    {'$oid': '5ff1e1df0a7214ada1000564'}              750.0   
17    {'$oid': '5ff1e1b40a7214ada100055b'}              750.0   
18    {'$oid': '5ff1e1eb0a720f0523000576'}              300.0   
19    {'$oid': '5ff1e1c80a720f052300056d'}                5.0   
20    {'$oid': '5f9c74f90a7214ad07000038'}              300.0   
21    {'$oid': '5ff1e1960a720f0523000567'}              750.0   
22    {'$oid': '5ff1e1ec0a7214ada100056c'}              150.0   
23    {'$oid': '5ff1e1d70a720f0523000571'}               40.0   
24    {'$oid': '5ff1e1c60a7214ada100055e'}                5.0   
25    {'$oid': '5ff1e1ee0a7214ada1000570'}                5.0   
26    {'$oid': '5ff1e1cf0a720f0523000570'}                5.0   
27    {'$oid': '5ff1e1e10a720f0523000572'}               25.0   
28    {'$oid': '5ff1e1d40a7214ada1000562'}                NaN   
29    {'$oid': '5ff1e1ee0a720f0523000578'}                5.0   
...                                    ...                ...   
1089  {'$oid': '603cbbb50a720fde100003e3'}               25.0   
1090  {'$oid': '603cadab0a720fde100003d6'}                NaN   
1091  {'$oid': '603c8cdf0a7217c72c0003bb'}                NaN   
1092  {'$oid': '603d30e60a7217c72c00043f'}               25.0   
1093  {'$oid': '603c65060a720fde10000395'}                NaN   
1094  {'$oid': '603d5d6c0a7217c72c000463'}               25.0   
1095  {'$oid': '603d760e0a720fde1000048e'}                NaN   
1096  {'$oid': '603d59e70a7217c72c00045f'}               25.0   
1097  {'$oid': '603c5ba30a720fde1000038d'}                NaN   
1098  {'$oid': '603c51160a7217c72c00038b'}                NaN   
1099  {'$oid': '603d40250a720fde10000459'}               25.0   
1100  {'$oid': '603c88240a7217c72c0003b8'}               25.0   
1101  {'$oid': '603cba880a7217c72c0003e0'}               25.0   
1102  {'$oid': '603cd4570a7217c72c0003f5'}                NaN   
1103  {'$oid': '603cd57d0a7217c72c0003f6'}               25.0   
1104  {'$oid': '603cf2ce0a7217c72c000413'}                NaN   
1105  {'$oid': '603cac7d0a7217c72c0003d7'}                NaN   
1106  {'$oid': '603c7c6c0a7217c72c0003b3'}               25.0   
1107  {'$oid': '603d28b60a720fde10000445'}                NaN   
1108  {'$oid': '603ca44a0a720fde100003d0'}                NaN   
1109  {'$oid': '603c4a0e0a720fde10000380'}                NaN   
1110  {'$oid': '603c6adf0a720fde1000039a'}                NaN   
1111  {'$oid': '603c9e6e0a720fde100003c7'}                NaN   
1112  {'$oid': '603c3d240a720fde10000373'}               25.0   
1113  {'$oid': '603cc2bc0a720fde100003e9'}               25.0   
1114  {'$oid': '603cc0630a720fde100003e6'}               25.0   
1115  {'$oid': '603d0b710a720fde1000042a'}                NaN   
1116  {'$oid': '603cf5290a720fde10000413'}                NaN   
1117  {'$oid': '603ce7100a7217c72c000405'}               25.0   
1118  {'$oid': '603c4fea0a7217c72c000389'}                NaN   

        

In [100]:
receipts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 15 columns):
_id                        1119 non-null object
bonusPointsEarned          544 non-null float64
bonusPointsEarnedReason    544 non-null object
createDate                 1119 non-null object
dateScanned                1119 non-null object
finishedDate               568 non-null object
modifyDate                 1119 non-null object
pointsAwardedDate          537 non-null object
pointsEarned               609 non-null float64
purchaseDate               671 non-null object
purchasedItemCount         635 non-null float64
rewardsReceiptItemList     679 non-null object
rewardsReceiptStatus       1119 non-null object
totalSpent                 684 non-null float64
userId                     1119 non-null object
dtypes: float64(4), object(11)
memory usage: 131.2+ KB


## Cleaning the "_id" from receipts dataset

In [101]:
#There are "_id" to be cleaned in multiple dataset, below is the function to clean the "_id" column

def clean_id(column, dataset):   
    variable = []
    for i in range(0, len(dataset)):
        for key, value in column[i].items():
            variable.append(value)
            
            
    return variable

In [102]:
receipts['_id'] = clean_id(receipts['_id'],receipts) #Clean id from receipts table 

## Cleaning the "dates" from receipts dataset

In [103]:
#There are multiple "date" column to be cleaned in multiple dataset, below is the function to clean the "date" column

def clean_date(column, dataset): #fuction to clean date in the dataset
    
    var = []
    
    for i in range(0,len(dataset)):
        if(type(column[i])) == dict:
            for key, value in column[i].items():
                var.append(value)
        else:
            var.append(None)
        
    return var

In [104]:
createDate = clean_date(receipts['createDate'],receipts)
receipts['createDate'] = pd.to_datetime(createDate, unit = 'ms')

dateScanned = clean_date(receipts['dateScanned'],receipts)
receipts['dateScanned'] = pd.to_datetime(dateScanned, unit = 'ms')

finishedDate = clean_date(receipts['finishedDate'],receipts)
receipts['finishedDate'] = pd.to_datetime(finishedDate, unit = 'ms')

modifyDate = clean_date(receipts['modifyDate'],receipts)
receipts['modifyDate'] = pd.to_datetime(modifyDate, unit = 'ms')

pointsAwardedDate = clean_date(receipts['pointsAwardedDate'],receipts)
receipts['pointsAwardedDate'] = pd.to_datetime(pointsAwardedDate, unit = 'ms')

purchaseDate = clean_date(receipts['purchaseDate'],receipts)
receipts['purchaseDate'] = pd.to_datetime(purchaseDate, unit = 'ms')

In [105]:
#Making all the column names in lower case in order to write sql query
receipts.columns= receipts.columns.str.lower()

In [106]:
receipts

_id  bonuspointsearned  \
0     5ff1e1eb0a720f0523000575              500.0   
1     5ff1e1bb0a720f052300056b              150.0   
2     5ff1e1f10a720f052300057a                5.0   
3     5ff1e1ee0a7214ada100056f                5.0   
4     5ff1e1d20a7214ada1000561                5.0   
5     5ff1e1e40a7214ada1000566              750.0   
6     5ff1e1cd0a720f052300056f                5.0   
7     5ff1e1a40a720f0523000569              500.0   
8     5ff1e1ed0a7214ada100056e                5.0   
9     5ff1e1eb0a7214ada100056b              250.0   
10    5ff1e1c50a720f052300056c              100.0   
11    5ff1e1a10a720f0523000568              750.0   
12    5ff1e1b60a7214ada100055c              150.0   
13    5f9c74f70a7214ad07000037              750.0   
14    5ff1e1b20a7214ada100055a              300.0   
15    5ff1e1e90a7214ada1000569                NaN   
16    5ff1e1df0a7214ada1000564              750.0   
17    5ff1e1b40a7214ada100055b              750.0   
18    5ff1e1eb0a720f0523000576              300.0   
19    5ff1e1c80a720f052300056d                5.0   
20    5f9c74f90a7214ad07000038              300.0   
21    5ff1e1960a720f0523000567              750.0   
22    5ff1e1ec0a7214ada100056c              150.0   
23    5ff1e1d70a720f0523000571               40.0   
24    5ff1e1c60a7214ada100055e                5.0   
25    5ff1e1ee0a7214ada1000570                5.0   
26    5ff1e1cf0a720f0523000570                5.0   
27    5ff1e1e10a720f0523000572               25.0   
28    5ff1e1d40a7214ada1000562                NaN   
29    5ff1e1ee0a720f0523000578                5.0   
...                        ...                ...   
1089  603cbbb50a720fde100003e3               25.0   
1090  603cadab0a720fde100003d6                NaN   
1091  603c8cdf0a7217c72c0003bb                NaN   
1092  603d30e60a7217c72c00043f               25.0   
1093  603c65060a720fde10000395                NaN   
1094  603d5d6c0a7217c72c000463               25.0   
1095  603d760e0a720fde1000048e                NaN   
1096  603d59e70a7217c72c00045f               25.0   
1097  603c5ba30a720fde1000038d                NaN   
1098  603c51160a7217c72c00038b                NaN   
1099  603d40250a720fde10000459               25.0   
1100  603c88240a7217c72c0003b8               25.0   
1101  603cba880a7217c72c0003e0               25.0   
1102  603cd4570a7217c72c0003f5                NaN   
1103  603cd57d0a7217c72c0003f6               25.0   
1104  603cf2ce0a7217c72c000413                NaN   
1105  603cac7d0a7217c72c0003d7                NaN   
1106  603c7c6c0a7217c72c0003b3               25.0   
1107  603d28b60a720fde10000445                NaN   
1108  603ca44a0a720fde100003d0                NaN   
1109  603c4a0e0a720fde10000380                NaN   
1110  603c6adf0a720fde1000039a                NaN   
1111  603c9e6e0a720fde100003c7                NaN   
1112  603c3d240a720fde10000373               25.0   
1113  603cc2bc0a720fde100003e9               25.0   
1114  603cc0630a720fde100003e6               25.0   
1115  603d0b710a720fde1000042a                NaN   
1116  603cf5290a720fde10000413                NaN   
1117  603ce7100a7217c72c000405               25.0   
1118  603c4fea0a7217c72c000389                NaN   

                                bonuspointsearnedreason  \
0     Receipt number 2 completed, bonus point schedu...   
1     Receipt number 5 completed, bonus point schedu...   
2                            All-receipts receipt bonus   
3                            All-receipts receipt bonus   
4                            All-receipts receipt bonus   
5     Receipt number 1 completed, bonus point schedu...   
6                            All-receipts receipt bonus   
7     Receipt number 2 completed, bonus point schedu...   
8                            All-receipts receipt bonus   
9     Receipt number 3 completed, bonus point schedu...   
10    Receipt number 6 completed, bonus point schedu...   
11    Receipt number 1 completed

## Creating  new table named "rewards_receipt" out of column "rewards_receiptItemList" from receipts dataset

In [107]:
receipts['rewardsreceiptitemlist']

0       [{'barcode': '4011', 'description': 'ITEM NOT ...
1       [{'barcode': '4011', 'description': 'ITEM NOT ...
2       [{'needsFetchReview': False, 'partnerItemId': ...
3       [{'barcode': '4011', 'description': 'ITEM NOT ...
4       [{'barcode': '4011', 'description': 'ITEM NOT ...
5       [{'barcode': '4011', 'description': 'ITEM NOT ...
6       [{'brandCode': 'MISSION', 'competitorRewardsGr...
7       [{'barcode': '046000832517', 'brandCode': 'BRA...
8       [{'barcode': '4011', 'description': 'ITEM NOT ...
9       [{'barcode': '4011', 'description': 'ITEM NOT ...
10      [{'barcode': '4011', 'description': 'ITEM NOT ...
11      [{'barcode': '013562300631', 'description': 'A...
12      [{'barcode': '034100573065', 'description': 'M...
13      [{'barcode': '075925306254', 'competitiveProdu...
14      [{'barcode': '4011', 'description': 'ITEM NOT ...
15      [{'needsFetchReview': True, 'needsFetchReviewR...
16      [{'barcode': '4011', 'description': 'ITEM NOT ...
17      [{'bar

## Taking a temporary df to create a subset of receipts dataset, so we can flatten the "rewardsreceiptitemlist" column. this will open up the dictionary.

In [108]:
df = receipts[['_id','rewardsreceiptitemlist']]

In [109]:
out = []
for n, row in df.iterrows():
    if (type(row['rewardsreceiptitemlist'])) == list:
        for item in row['rewardsreceiptitemlist']:
            row['flatten'] = item
            out += [row.copy()]
    else:
        out+= [row.copy()]


flattened_df = pd.DataFrame(out)

In [110]:
flattened_df

#From the below output I can see that there are list of dictionaries in rewardsreceiptitemlist column 
#(Example: Interpreting that "_id" for index 1 has two dictionaries)

_id  \
0     5ff1e1eb0a720f0523000575   
1     5ff1e1bb0a720f052300056b   
1     5ff1e1bb0a720f052300056b   
2     5ff1e1f10a720f052300057a   
3     5ff1e1ee0a7214ada100056f   
4     5ff1e1d20a7214ada1000561   
4     5ff1e1d20a7214ada1000561   
5     5ff1e1e40a7214ada1000566   
6     5ff1e1cd0a720f052300056f   
7     5ff1e1a40a720f0523000569   
8     5ff1e1ed0a7214ada100056e   
9     5ff1e1eb0a7214ada100056b   
10    5ff1e1c50a720f052300056c   
11    5ff1e1a10a720f0523000568   
12    5ff1e1b60a7214ada100055c   
12    5ff1e1b60a7214ada100055c   
12    5ff1e1b60a7214ada100055c   
12    5ff1e1b60a7214ada100055c   
12    5ff1e1b60a7214ada100055c   
12    5ff1e1b60a7214ada100055c   
12    5ff1e1b60a7214ada100055c   
12    5ff1e1b60a7214ada100055c   
12    5ff1e1b60a7214ada100055c   
12    5ff1e1b60a7214ada100055c   
13    5f9c74f70a7214ad07000037   
13    5f9c74f70a7214ad07000037   
13    5f9c74f70a7214ad07000037   
13    5f9c74f70a7214ad07000037   
13    5f9c74f70a7214ad07000037   
13    5f9c74f70a7214ad07000037   
...                        ...   
1098  603c51160a7217c72c00038b   
1099  603d40250a720fde10000459   
1099  603d40250a720fde10000459   
1100  603c88240a7217c72c0003b8   
1100  603c88240a7217c72c0003b8   
1101  603cba880a7217c72c0003e0   
1101  603cba880a7217c72c0003e0   
1102  603cd4570a7217c72c0003f5   
1103  603cd57d0a7217c72c0003f6   
1103  603cd57d0a7217c72c0003f6   
1104  603cf2ce0a7217c72c000413   
1105  603cac7d0a7217c72c0003d7   
1106  603c7c6c0a7217c72c0003b3   
1106  603c7c6c0a7217c72c0003b3   
1107  603d28b60a720fde10000445   
1108  603ca44a0a720fde100003d0   
1109  603c4a0e0a720fde10000380   
1110  603c6adf0a720fde1000039a   
1111  603c9e6e0a720fde100003c7   
1112  603c3d240a720fde10000373   
1112  603c3d240a720fde10000373   
1113  603cc2bc0a720fde100003e9   
1113  603cc2bc0a720fde100003e9   
1114  603cc0630a720fde100003e6   
1114  603cc0630a720fde100003e6   
1115  603d0b710a720fde1000042a   
1116  603cf5290a720fde10000413   
1117  603ce7100a7217c72c000405   
1117  603ce7100a7217c72c000405   
1118  603c4fea0a7217c72c000389   

                                 rewardsreceiptitemlist  \
0     [{'barcode': '4011', 'description': 'ITEM NOT ...   
1     [{'barcode': '4011', 'description': 'ITEM NOT ...   
1     [{'barcode': '4011', 'description': 'ITEM NOT ...   
2     [{'needsFetchReview': False, 'partnerItemId': ...   
3     [{'barcode': '4011', 'description': 'ITEM NOT ...   
4     [{'barcode': '4011', 'description': 'ITEM NOT ...   
4     [{'barcode': '4011', 'description': 'ITEM NOT ...   
5     [{'barcode': '4011', 'description': 'ITEM NOT ...   
6     [{'brandCode': 'MISSION', 'competitorRewardsGr...   
7     [{'barcode': '046000832517', 'brandCode': 'BRA...   
8     [{'barcode': '4011', 'description': 'ITEM NOT ...   
9     [{'barcode': '4011', 'description': 'ITEM NOT ...   
10    [{'barcode': '4011', 'description': 'ITEM NOT ...   
11    [{'barcode': '013562300631', 'description': 'A...   
12    [{'barcode': '034100573065', 'description': 'M...   
12    [{'barcode': '034100573065', 'description': 'M...   
12    [{'barcode': '034100573065', 'description': 'M...   
12    [{'barcode': '034100573065', 'description': 'M...   
12    [{'barcode': '034100573065', 'description': 'M...   
12    [{'barcode': '034100573065', 'description': 'M...   
12    [{'barcode': '034100573065', 'description': 'M...   
12    [{'barcode': '034100573065', 'description': 'M...   
12    [{'barcode': '034100573065', 'description': 'M...   
12    [{'barcode': '034100573065', 'description': 'M...   
13    [{'barcode': '075925306254', 'competitiveProdu...   
13    [{'barcode': '075925306254', 'competitiveProdu...   
13    [{'barcode': '075925306254', 'competitiveProdu...   
13    [{'barcode': '075925306254', 'competitiveProdu...   
13    [{'barcode': '075925306254', 'competitiveProdu...   
13    [{'barcode': '075925306254', 'competitiveProdu...   
...                                                 ...   
1098                          

In [111]:
receipts = pd.merge(receipts,flattened_df, on = '_id') #Merging the flattened_df to my original receipts

In [112]:
#Since we have flattened_df we no longer need "rewardsreceiptitemlist", so dropping it from the dataset
receipts.drop('rewardsreceiptitemlist_x', inplace = True, axis = 1)
receipts.drop('rewardsreceiptitemlist_y', inplace = True, axis = 1)

In [113]:
#Renaming flatten as "rewardsreceiptitemlist" 
receipts.rename(columns = {'flatten':'rewardsreceiptitemlist'}, inplace = True)

In [114]:
#The code below will provide unique list of keys in all the dictionary (This is our column for the new table to be created)  

all_keys = set()

for i in range(len(receipts)):
    if (type(receipts['rewardsreceiptitemlist'][i])) == dict:
        all_keys = all_keys.union(set(receipts['rewardsreceiptitemlist'][i].keys()))

In [115]:
all_keys #These are all the unique keys from the 'rewardsreceiptitemlist' column

{'barcode',
 'brandCode',
 'competitiveProduct',
 'competitorRewardsGroup',
 'deleted',
 'description',
 'discountedItemPrice',
 'finalPrice',
 'itemNumber',
 'itemPrice',
 'metabriteCampaignId',
 'needsFetchReview',
 'needsFetchReviewReason',
 'originalFinalPrice',
 'originalMetaBriteBarcode',
 'originalMetaBriteDescription',
 'originalMetaBriteItemPrice',
 'originalMetaBriteQuantityPurchased',
 'originalReceiptItemText',
 'partnerItemId',
 'pointsEarned',
 'pointsNotAwardedReason',
 'pointsPayerId',
 'preventTargetGapPoints',
 'priceAfterCoupon',
 'quantityPurchased',
 'rewardsGroup',
 'rewardsProductPartnerId',
 'targetPrice',
 'userFlaggedBarcode',
 'userFlaggedDescription',
 'userFlaggedNewItem',
 'userFlaggedPrice',
 'userFlaggedQuantity'}

In [116]:
rewards_receipt = pd.DataFrame(columns = ['_id']+list(all_keys))  #Structure of new table to be created
rewards_receipt

Empty DataFrame
Columns: [_id, finalPrice, priceAfterCoupon, preventTargetGapPoints, originalMetaBriteBarcode, userFlaggedNewItem, needsFetchReviewReason, rewardsProductPartnerId, originalFinalPrice, needsFetchReview, rewardsGroup, originalMetaBriteDescription, pointsPayerId, originalMetaBriteItemPrice, discountedItemPrice, description, userFlaggedBarcode, competitiveProduct, userFlaggedDescription, partnerItemId, pointsEarned, targetPrice, metabriteCampaignId, deleted, quantityPurchased, barcode, competitorRewardsGroup, userFlaggedPrice, brandCode, itemPrice, userFlaggedQuantity, originalMetaBriteQuantityPurchased, originalReceiptItemText, itemNumber, pointsNotAwardedReason]
Index: []

[0 rows x 35 columns]

In [117]:
# Giving all the value to the respected key to the new structured table. Null will be assigned to the key if they dont have values

columns = ['_id']+list(all_keys)
dictionary = dict()


for each in columns:
    dictionary[each] = []
    
for i in range(len(receipts)):
    dictionary['_id'].append(receipts['_id'].values[0])
    tempDic = receipts['rewardsreceiptitemlist'][i]
    
    for each in all_keys:
        if (type(tempDic)) == dict:
            dictionary[each].append(tempDic.get(each,np.nan)) 
        else:
            dictionary[each].append(np.nan)

In [118]:
rewards_receipt = pd.DataFrame(dictionary)
rewards_receipt

_id finalPrice priceAfterCoupon  \
0     5ff1e1eb0a720f0523000575      26.00              NaN   
1     5ff1e1eb0a720f0523000575          1              NaN   
2     5ff1e1eb0a720f0523000575      10.00              NaN   
3     5ff1e1eb0a720f0523000575        NaN              NaN   
4     5ff1e1eb0a720f0523000575      28.00              NaN   
5     5ff1e1eb0a720f0523000575          1              NaN   
6     5ff1e1eb0a720f0523000575       2.56              NaN   
7     5ff1e1eb0a720f0523000575       3.25              NaN   
8     5ff1e1eb0a720f0523000575       2.23              NaN   
9     5ff1e1eb0a720f0523000575      10.00              NaN   
10    5ff1e1eb0a720f0523000575      20.00              NaN   
11    5ff1e1eb0a720f0523000575      20.00              NaN   
12    5ff1e1eb0a720f0523000575          1              NaN   
13    5ff1e1eb0a720f0523000575      50.00              NaN   
14    5ff1e1eb0a720f0523000575         29              NaN   
15    5ff1e1eb0a720f0523000575         29              NaN   
16    5ff1e1eb0a720f0523000575         29              NaN   
17    5ff1e1eb0a720f0523000575         29              NaN   
18    5ff1e1eb0a720f0523000575         29              NaN   
19    5ff1e1eb0a720f0523000575         29              NaN   
20    5ff1e1eb0a720f0523000575         29              NaN   
21    5ff1e1eb0a720f0523000575         29              NaN   
22    5ff1e1eb0a720f0523000575         29              NaN   
23    5ff1e1eb0a720f0523000575         29              NaN   
24    5ff1e1eb0a720f0523000575          1              NaN   
25    5ff1e1eb0a720f0523000575        NaN              NaN   
26    5ff1e1eb0a720f0523000575        NaN              NaN   
27    5ff1e1eb0a720f0523000575        NaN              NaN   
28    5ff1e1eb0a720f0523000575        NaN              NaN   
29    5ff1e1eb0a720f0523000575        NaN              NaN   
...                        ...        ...              ...   
7351  5ff1e1eb0a720f0523000575        NaN              NaN   
7352  5ff1e1eb0a720f0523000575      22.97            22.97   
7353  5ff1e1eb0a720f0523000575      11.99            11.99   
7354  5ff1e1eb0a720f0523000575      22.97            22.97   
7355  5ff1e1eb0a720f0523000575      11.99            11.99   
7356  5ff1e1eb0a720f0523000575      22.97            22.97   
7357  5ff1e1eb0a720f0523000575      11.99            11.99   
7358  5ff1e1eb0a720f0523000575        NaN              NaN   
7359  5ff1e1eb0a720f0523000575      22.97            22.97   
7360  5ff1e1eb0a720f0523000575      11.99            11.99   
7361  5ff1e1eb0a720f0523000575        NaN              NaN   
7362  5ff1e1eb0a720f0523000575        NaN              NaN   
7363  5ff1e1eb0a720f0523000575      22.97            22.97   
7364  5ff1e1eb0a720f0523000575      11.99            11.99   
7365  5ff1e1eb0a720f0523000575        NaN              NaN   
7366  5ff1e1eb0a720f0523000575        NaN              NaN   
7367  5ff1e1eb0a720f0523000575        NaN              NaN   
7368  5ff1e1eb0a720f0523000575        NaN              NaN   
7369  5ff1e1eb0a720f0523000575        NaN              NaN   
7370  5ff1e1eb0a720f0523000575      22.97            22.97   
7371  5ff1e1eb0a720f0523000575      11.99            11.99   
7372  5ff1e1eb0a720f0523000575      22.97            22.97   
7373  5ff1e1eb0a720f0523000575      11.99            11.99   
7374  5ff1e1eb0a720f0523000575      22.97            22.97   
7375  5ff1e1eb0a720f0523000575      11.99            11.99   
7376  5ff1e1eb0a720f0523000575        NaN              NaN   
7377  5ff1e1eb0a720f0523000575        NaN              NaN   
7378  5ff1e1eb0a720f0523000575      22.97            22.97   
7379  5ff1e1eb0a720f0523000575      11.99            11.99   
7380  5ff1e1eb0a720f0523000575        NaN              NaN   

     preventTargetGapPoints originalMetaBriteBarcode userFlaggedNewItem  \
0                      True                      NaN               True   
1                       NaN     

In [119]:
#Converting all the column names in lower case in order to write sql query
rewards_receipt.columns= rewards_receipt.columns.str.lower()

In [120]:
#"rewardsreceiptitemlist" column is no longer needed in the original receipts dataset
#Since a new table rewards_receipt was created from this rewardsreceiptitemlist column

receipts.drop('rewardsreceiptitemlist', inplace = True, axis = 1)

In [121]:
receipts

_id  bonuspointsearned  \
0     5ff1e1eb0a720f0523000575              500.0   
1     5ff1e1bb0a720f052300056b              150.0   
2     5ff1e1bb0a720f052300056b              150.0   
3     5ff1e1f10a720f052300057a                5.0   
4     5ff1e1ee0a7214ada100056f                5.0   
5     5ff1e1d20a7214ada1000561                5.0   
6     5ff1e1d20a7214ada1000561                5.0   
7     5ff1e1e40a7214ada1000566              750.0   
8     5ff1e1cd0a720f052300056f                5.0   
9     5ff1e1a40a720f0523000569              500.0   
10    5ff1e1ed0a7214ada100056e                5.0   
11    5ff1e1eb0a7214ada100056b              250.0   
12    5ff1e1c50a720f052300056c              100.0   
13    5ff1e1a10a720f0523000568              750.0   
14    5ff1e1b60a7214ada100055c              150.0   
15    5ff1e1b60a7214ada100055c              150.0   
16    5ff1e1b60a7214ada100055c              150.0   
17    5ff1e1b60a7214ada100055c              150.0   
18    5ff1e1b60a7214ada100055c              150.0   
19    5ff1e1b60a7214ada100055c              150.0   
20    5ff1e1b60a7214ada100055c              150.0   
21    5ff1e1b60a7214ada100055c              150.0   
22    5ff1e1b60a7214ada100055c              150.0   
23    5ff1e1b60a7214ada100055c              150.0   
24    5f9c74f70a7214ad07000037              750.0   
25    5f9c74f70a7214ad07000037              750.0   
26    5f9c74f70a7214ad07000037              750.0   
27    5f9c74f70a7214ad07000037              750.0   
28    5f9c74f70a7214ad07000037              750.0   
29    5f9c74f70a7214ad07000037              750.0   
...                        ...                ...   
7351  603c51160a7217c72c00038b                NaN   
7352  603d40250a720fde10000459               25.0   
7353  603d40250a720fde10000459               25.0   
7354  603c88240a7217c72c0003b8               25.0   
7355  603c88240a7217c72c0003b8               25.0   
7356  603cba880a7217c72c0003e0               25.0   
7357  603cba880a7217c72c0003e0               25.0   
7358  603cd4570a7217c72c0003f5                NaN   
7359  603cd57d0a7217c72c0003f6               25.0   
7360  603cd57d0a7217c72c0003f6               25.0   
7361  603cf2ce0a7217c72c000413                NaN   
7362  603cac7d0a7217c72c0003d7                NaN   
7363  603c7c6c0a7217c72c0003b3               25.0   
7364  603c7c6c0a7217c72c0003b3               25.0   
7365  603d28b60a720fde10000445                NaN   
7366  603ca44a0a720fde100003d0                NaN   
7367  603c4a0e0a720fde10000380                NaN   
7368  603c6adf0a720fde1000039a                NaN   
7369  603c9e6e0a720fde100003c7                NaN   
7370  603c3d240a720fde10000373               25.0   
7371  603c3d240a720fde10000373               25.0   
7372  603cc2bc0a720fde100003e9               25.0   
7373  603cc2bc0a720fde100003e9               25.0   
7374  603cc0630a720fde100003e6               25.0   
7375  603cc0630a720fde100003e6               25.0   
7376  603d0b710a720fde1000042a                NaN   
7377  603cf5290a720fde10000413                NaN   
7378  603ce7100a7217c72c000405               25.0   
7379  603ce7100a7217c72c000405               25.0   
7380  603c4fea0a7217c72c000389                NaN   

                                bonuspointsearnedreason  \
0     Receipt number 2 completed, bonus point schedu...   
1     Receipt number 5 completed, bonus point schedu...   
2     Receipt number 5 completed, bonus point schedu...   
3                            All-receipts receipt bonus   
4                            All-receipts receipt bonus   
5                            All-receipts receipt bonus   
6                            All-receipts receipt bonus   
7     Receipt number 1 completed, bonus point schedu...   
8                            All-receipts receipt bonus   
9     Receipt number 2 completed, bonus point schedu...   
10                           All-receipts receipt bonus   
11    Receipt number 3 completed

## Second Dataset (users)

In [122]:
#Reading code for the second dataset

users = pd.read_json('./users.json', lines = True)
users

_id  active               createdDate  \
0    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3    {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
5    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
6    {'$oid': '5ff1e1e8cfcf6c399c274ad9'}    True  {'$date': 1609687528354}   
7    {'$oid': '5ff1e1b7cfcf6c399c274a5a'}    True  {'$date': 1609687479626}   
8    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
9    {'$oid': '5ff1e1f1cfcf6c399c274b0b'}    True  {'$date': 1609687537564}   
10   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
11   {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
12   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
13   {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
14   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
15   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
16   {'$oid': '5ff1e1e4cfcf6c399c274ac3'}    True  {'$date': 1609687524656}   
17   {'$oid': '5ff1e1b4cfcf6c399c274a54'}    True  {'$date': 1609687476410}   
18   {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
19   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
20   {'$oid': '5ff370c562fde912123a5e0e'}    True  {'$date': 1609789637776}   
21   {'$oid': '5ff36d0362fde912123a5535'}    True  {'$date': 1609788675973}   
22   {'$oid': '5ff36d83135e7011bcb864d6'}    True  {'$date': 1609788803244}   
23   {'$oid': '5ff36c8862fde912123a538a'}    True  {'$date': 1609788552943}   
24   {'$oid': '5ff370c562fde912123a5e0e'}    True  {'$date': 1609789637776}   
25   {'$oid': '5ff36be7135e7011bcb856d3'}    True  {'$date': 1609788391239}   
26   {'$oid': '5ff36d0362fde912123a5535'}    True  {'$date': 1609788675973}   
27   {'$oid': '5ff36be7135e7011bcb856d3'}    True  {'$date': 1609788391239}   
28   {'$oid': '5ff36a3862fde912123a4460'}    True  {'$date': 1609787960352}   
29   {'$oid': '5ff36d0362fde912123a5535'}    True  {'$date': 1609788675973}   
..                                    ...     ...                       ...   
465  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
466  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
467  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
468  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
469  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
470  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
471  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
472  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
473  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
474  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
475  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
476  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
477  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
478  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
479  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
480  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
481  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
482  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
483  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
484  {'$o

In [123]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
_id             495 non-null object
active          495 non-null bool
createdDate     495 non-null object
lastLogin       433 non-null object
role            495 non-null object
signUpSource    447 non-null object
state           439 non-null object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


In [124]:
#Making all the column names in lower case in order to write sql query
users.columns= users.columns.str.lower()

In [125]:
#Using the "clean_id" function to clean the _id in users dataset
users['_id'] = clean_id(users['_id'],users)

In [126]:
#Using the "clean_date" function to clean the dates in users dataset

createdDate = clean_date(users['createddate'],users)
users['createddate'] = pd.to_datetime(createdDate, unit = 'ms')

lastLogin = clean_date(users['lastlogin'],users)
users['lastlogin'] = pd.to_datetime(lastLogin, unit = 'ms')


In [127]:
users #Cleaned second dataset

_id  active             createddate  \
0    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
1    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
2    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
3    5ff1e1eacfcf6c399c274ae6    True 2021-01-03 15:25:30.554   
4    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
5    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
6    5ff1e1e8cfcf6c399c274ad9    True 2021-01-03 15:25:28.354   
7    5ff1e1b7cfcf6c399c274a5a    True 2021-01-03 15:24:39.626   
8    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
9    5ff1e1f1cfcf6c399c274b0b    True 2021-01-03 15:25:37.564   
10   5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
11   5ff1e1eacfcf6c399c274ae6    True 2021-01-03 15:25:30.554   
12   5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
13   5ff1e1eacfcf6c399c274ae6    True 2021-01-03 15:25:30.554   
14   5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
15   5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
16   5ff1e1e4cfcf6c399c274ac3    True 2021-01-03 15:25:24.656   
17   5ff1e1b4cfcf6c399c274a54    True 2021-01-03 15:24:36.410   
18   5ff1e1eacfcf6c399c274ae6    True 2021-01-03 15:25:30.554   
19   5ff1e194b6a9d73a3a9f1052    True 2021-01-03 15:24:04.800   
20   5ff370c562fde912123a5e0e    True 2021-01-04 19:47:17.776   
21   5ff36d0362fde912123a5535    True 2021-01-04 19:31:15.973   
22   5ff36d83135e7011bcb864d6    True 2021-01-04 19:33:23.244   
23   5ff36c8862fde912123a538a    True 2021-01-04 19:29:12.943   
24   5ff370c562fde912123a5e0e    True 2021-01-04 19:47:17.776   
25   5ff36be7135e7011bcb856d3    True 2021-01-04 19:26:31.239   
26   5ff36d0362fde912123a5535    True 2021-01-04 19:31:15.973   
27   5ff36be7135e7011bcb856d3    True 2021-01-04 19:26:31.239   
28   5ff36a3862fde912123a4460    True 2021-01-04 19:19:20.352   
29   5ff36d0362fde912123a5535    True 2021-01-04 19:31:15.973   
..                        ...     ...                     ...   
465  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
466  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
467  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
468  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
469  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
470  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
471  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
472  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
473  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
474  5fa41775898c7a11a6bcef3e    True 2020-11-05 15:17:09.396   
475  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
476  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
477  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
478  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
479  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
480  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
481  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
482  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
483  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
484  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
485  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
486  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
487  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
488  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
489  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
490  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
491  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
492  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
493  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381   
494  54943462e4b07e684157a532    True 2014-12-19 14:21:22.381

## Third Dataset (brands)

In [128]:
#Reading the third dataset

brands = pd.read_json('./brands.json', lines = True)
brands

_id       barcode  \
0     {'$oid': '601ac115be37ce2ead437551'}  511111019862   
1     {'$oid': '601c5460be37ce2ead43755f'}  511111519928   
2     {'$oid': '601ac142be37ce2ead43755d'}  511111819905   
3     {'$oid': '601ac142be37ce2ead43755a'}  511111519874   
4     {'$oid': '601ac142be37ce2ead43755e'}  511111319917   
5     {'$oid': '601ac142be37ce2ead43755b'}  511111719885   
6     {'$oid': '601ac142be37ce2ead43755c'}  511111219897   
7     {'$oid': '5cdad0f5166eb33eb7ce0faa'}  511111104810   
8     {'$oid': '5ab15636e4b0be0a89bb0b07'}  511111504412   
9     {'$oid': '5c408e8bcd244a1fdb47aee7'}  511111504788   
10    {'$oid': '5f4bf556be37ce0b4491554d'}  511111516354   
11    {'$oid': '57c08106e4b0718ff5fcb02c'}  511111102540   
12    {'$oid': '588ba07be4b02187f85cdadd'}  511111201076   
13    {'$oid': '5d6413156d5f3b23d1bc790a'}  511111205012   
14    {'$oid': '585a9611e4b03e62d1ce0e74'}  511111801801   
15    {'$oid': '57e5820ce4b0ac389136a311'}  511111202233   
16    {'$oid': '5fb6adb8be37ce522e165cb8'}  511111817376   
17    {'$oid': '5f358338be37ce443bf9d55a'}  511111515319   
18    {'$oid': '5fb28549be37ce522e165cb5'}  511111317364   
19    {'$oid': '592486bfe410d61fcea3d139'}  511111300700   
20    {'$oid': '5c4699f387ff3577e203ea29'}  511111305125   
21    {'$oid': '5da6071ea60b87376833e34d'}  511111005650   
22    {'$oid': '57ebc011e4b0ac389136a335'}  511111802129   
23    {'$oid': '5332f5fee4b03c9a25efd0bd'}  511111303947   
24    {'$oid': '5332fa7ce4b03c9a25efd22e'}  511111802914   
25    {'$oid': '5e9f18bfbe37ce3e45b6a77f'}  511111914549   
26    {'$oid': '592486bee410d61fcea3d133'}  511111400769   
27    {'$oid': '5d66d71fa3a018093ab34728'}  511111105329   
28    {'$oid': '5f493e72be37ce64d0ae36c6'}  511111316114   
29    {'$oid': '5f4936ddbe37ce52f8314fd9'}  511111315957   
...                                    ...           ...   
1137  {'$oid': '57ebbfa9e4b0ac389136a331'}  511111202165   
1138  {'$oid': '5a5d2c59e4b06ba572cf24a9'}  511111900054   
1139  {'$oid': '5887a051e4b02187f85cdad0'}  511111401209   
1140  {'$oid': '5ab1528ce4b0915382df761c'}  511111504344   
1141  {'$oid': '5d66d6016d5f3b6188d4f047'}  511111405306   
1142  {'$oid': '5f21c54abe37ce7958c59528'}  511111315230   
1143  {'$oid': '5332f7fde4b03c9a25efd168'}  511111903345   
1144  {'$oid': '57c08242e4b0718ff5fcb032'}  511111202516   
1145  {'$oid': '5f5bc4f1be37ce17125ac0e9'}  511111716594   
1146  {'$oid': '5332fa12e4b03c9a25efd1e6'}  511111703105   
1147  {'$oid': '5dc0351a1dda2c0ad7da64aa'}  511111306238   
1148  {'$oid': '5d66e2d46d5f3b6188d4f056'}  511111905523   
1149  {'$oid': '5a0219ede4b00d6a5ccb394e'}  511111400554   
1150  {'$oid': '5d66da306d5f3b6188d4f04c'}  511111005377   
1151  {'$oid': '5d66dfe6a3a018093ab3472c'}  511111905479   
1152  {'$oid': '5f3e9172be37ce5a0e01b955'}  511111715559   
1153  {'$oid': '58b598dbe4b0857c2ddb7253'}  511111101017   
1154  {'$oid': '57d95826e4b0ac389136a2b7'}  511111302346   
1155  {'$oid': '5a5d1bffe4b0db471c2d0423'}  511111000396   
1156  {'$oid': '5fb82fe3be37ce522e165cea'}  511111617853   
1157  {'$oid': '5332fa75e4b03c9a25efd221'}  511111303015   
1158  {'$oid': '5f628215be37ce78e6e2efab'}  511111716648   
1159  {'$oid': '585a96cbe4b03e62d1ce0e88'}  511111501619   
1160  {'$oid': '5887a216e4b02187f85cdad5'}  511111401155   
1161  {'$oid': '5332f709e4b03c9a25efd0f2'}  511111403845   
1162  {'$oid': '5f77274dbe37ce6b592e90c0'}  511111116752   
1163  {'$oid': '5dc1fca91dda2c0ad7da64ae'}  511111706328   
1164  {'$oid': '5f494c6e04db711dd8fe87e7'}  511111416173   
1165  {'$oid': '5a021611e4b00efe02b02a57'}  511111400608   
1166  {'$oid': '6026d757be37ce6369301468'}  511111019930   

                          brandCode                     category  \
0                               NaN                       Baking   
1                         STARBUCKS                    Beverages   
2     TEST BRANDCODE @1612366146176                       Baking   
3     TEST BRANDCODE @1612366146051           

In [136]:
#In the cpg column, I observed that there is an "oid" inside dictionary. 
#I wanted to check if this is same as "_id" in receipts data set
#I wanted to check if this is same as "_id" in users data set

new = brands['cpg']

oid = set()
for i, d in new.iteritems():
    oid.add(d['$id']['$oid'])

oid

{'5332f5ebe4b03c9a25efd0a8',
 '5332f5f2e4b03c9a25efd0aa',
 '5332f5f3e4b03c9a25efd0ac',
 '5332f5f3e4b03c9a25efd0ae',
 '5332f5f6e4b03c9a25efd0b4',
 '5332f5fbe4b03c9a25efd0ba',
 '5332f6ede4b03c9a25efd0e6',
 '5332f709e4b03c9a25efd0f1',
 '5332f712e4b03c9a25efd0f9',
 '5332f734e4b03c9a25efd104',
 '5332f7a7e4b03c9a25efd134',
 '5332f7ffe4b03c9a25efd16b',
 '5332fa12e4b03c9a25efd1e7',
 '5332fa58e4b03c9a25efd215',
 '53501d2ab9237fd7718f9e18',
 '53501d2ab9237fd7718f9e1a',
 '537b7ec2892471241348d709',
 '53e10d6368abd3c7065097cc',
 '540f215de4b09fb9af4b49f1',
 '5459429be4b0bfcb1e864082',
 '550b2565e4b001d5e9e4146f',
 '559c2234e4b06aca36af13c6',
 '55a41a91e4b0d0a65b3692ec',
 '55a41b4be4b0d0a65b3692ee',
 '55a58719e4b0d0a65b3693cc',
 '55b62995e4b0d8e685c14213',
 '56462469e4b0c05761111936',
 '57ace90fe4b057c992a316ab',
 '581ce8d5e4b058c71f74c442',
 '59b96571e4b0bf7c7980588b',
 '59ba6f1ce4b092b29c167346',
 '5a734034e4b0d58f376be874',
 '5c45f8b087ff3552f950f026',
 '5c4f20b01b4181298aacffc1',
 '5c6c2ee99514

In [138]:
(oid).intersection(set(receipts['_id'])) 

#There is no common output which means both the column are different from receipts and brands dataset

set()

In [139]:
(oid).intersection(set(users['_id'])) 

#There is no common output which means both the column are different from receipts and users dataset

set()

In [140]:
#Above was a test to check if there was an intersecting between the values in the cpg column and the other tables 
#But no relationship was established and also there was no clear understanding of the cpg column and its use in further analysis
#Therefore dropping the column 

In [129]:
brands.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 8 columns):
_id             1167 non-null object
barcode         1167 non-null int64
brandCode       933 non-null object
category        1012 non-null object
categoryCode    517 non-null object
cpg             1167 non-null object
name            1167 non-null object
topBrand        555 non-null float64
dtypes: float64(1), int64(1), object(6)
memory usage: 73.0+ KB


In [130]:
#Making all the column names in lower case in order to write sql query
brands.columns= brands.columns.str.lower()

In [131]:
#Using the "clean_id" function to clean the _id in brands dataset
brands['_id'] = clean_id(brands['_id'],brands)

In [132]:
brands

_id       barcode                      brandcode  \
0     601ac115be37ce2ead437551  511111019862                            NaN   
1     601c5460be37ce2ead43755f  511111519928                      STARBUCKS   
2     601ac142be37ce2ead43755d  511111819905  TEST BRANDCODE @1612366146176   
3     601ac142be37ce2ead43755a  511111519874  TEST BRANDCODE @1612366146051   
4     601ac142be37ce2ead43755e  511111319917  TEST BRANDCODE @1612366146827   
5     601ac142be37ce2ead43755b  511111719885  TEST BRANDCODE @1612366146091   
6     601ac142be37ce2ead43755c  511111219897  TEST BRANDCODE @1612366146133   
7     5cdad0f5166eb33eb7ce0faa  511111104810                     J.L. KRAFT   
8     5ab15636e4b0be0a89bb0b07  511111504412           CAMPBELLS HOME STYLE   
9     5c408e8bcd244a1fdb47aee7  511111504788                           TEST   
10    5f4bf556be37ce0b4491554d  511111516354  TEST BRANDCODE @1598813526777   
11    57c08106e4b0718ff5fcb02c  511111102540                            NaN   
12    588ba07be4b02187f85cdadd  511111201076                        CALUMET   
13    5d6413156d5f3b23d1bc790a  511111205012                   511111205012   
14    585a9611e4b03e62d1ce0e74  511111801801              AUNT JEMIMA SYRUP   
15    57e5820ce4b0ac389136a311  511111202233                         MOLSON   
16    5fb6adb8be37ce522e165cb8  511111817376                       LOTRIMIN   
17    5f358338be37ce443bf9d55a  511111515319  TEST BRANDCODE @1597342520277   
18    5fb28549be37ce522e165cb5  511111317364                            NaN   
19    592486bfe410d61fcea3d139  511111300700                        ST IVES   
20    5c4699f387ff3577e203ea29  511111305125                     CHRISIMAGE   
21    5da6071ea60b87376833e34d  511111005650                   ALKA SELTZER   
22    57ebc011e4b0ac389136a335  511111802129         JACK DANIEL'S BARBECUE   
23    5332f5fee4b03c9a25efd0bd  511111303947                            NaN   
24    5332fa7ce4b03c9a25efd22e  511111802914                            NaN   
25    5e9f18bfbe37ce3e45b6a77f  511111914549                            NaN   
26    592486bee410d61fcea3d133  511111400769               MAGNUM Ice Cream   
27    5d66d71fa3a018093ab34728  511111105329                   511111105329   
28    5f493e72be37ce64d0ae36c6  511111316114  TEST BRANDCODE @1598635634882   
29    5f4936ddbe37ce52f8314fd9  511111315957                            NaN   
...                        ...           ...                            ...   
1137  57ebbfa9e4b0ac389136a331  511111202165                                  
1138  5a5d2c59e4b06ba572cf24a9  511111900054                        SUNSILK   
1139  5887a051e4b02187f85cdad0  511111401209                        BAKER'S   
1140  5ab1528ce4b0915382df761c  511111504344                     ECCE PANIS   
1141  5d66d6016d5f3b6188d4f047  511111405306                   511111405306   
1142  5f21c54abe37ce7958c59528  511111315230                       AIR WICK   
1143  5332f7fde4b03c9a25efd168  511111903345                         HAMM'S   
1144  57c08242e4b0718ff5fcb032  511111202516                            NaN   
1145  5f5bc4f1be37ce17125ac0e9  511111716594  TEST BRANDCODE @1599849713740   
1146  5332fa12e4b03c9a25efd1e6  511111703105                            NaN   
1147  5dc0351a1dda2c0ad7da64aa  511111306238                MONSTERS CEREAL   
1148  5d66e2d46d5f3b6188d4f056  511111905523              TAZO BOTTLED TEAS   
1149  5a0219ede4b00d6a5ccb394e  511111400554                       TONI&GUY   
1150  5d66da306d5f3b6188d4f04c  511111005377                   511111005377   
1151  5d66dfe6a3a018093ab3472c  511111905479                   511111905479   
1152  5f3e9172be37ce5a0e01b955  511111715559  TEST BRANDCODE @1597935986434   
1153  58b598dbe4b0857c2ddb7253  511111101017                       MICKEY'S   
1154  57d95826e4b0ac389136a2b7  511111302346                  CRYSTAL LIGHT   
1155  5a5d1bffe4b0db471c2d0423  511111000396               COORS EXTRA GOLD  

In [86]:
brands.drop('cpg', inplace = True, axis = 1) #Dropping the cpg column

In [87]:
brands #Cleaned brands dataset

_id       barcode                      brandcode  \
0     601ac115be37ce2ead437551  511111019862                            NaN   
1     601c5460be37ce2ead43755f  511111519928                      STARBUCKS   
2     601ac142be37ce2ead43755d  511111819905  TEST BRANDCODE @1612366146176   
3     601ac142be37ce2ead43755a  511111519874  TEST BRANDCODE @1612366146051   
4     601ac142be37ce2ead43755e  511111319917  TEST BRANDCODE @1612366146827   
5     601ac142be37ce2ead43755b  511111719885  TEST BRANDCODE @1612366146091   
6     601ac142be37ce2ead43755c  511111219897  TEST BRANDCODE @1612366146133   
7     5cdad0f5166eb33eb7ce0faa  511111104810                     J.L. KRAFT   
8     5ab15636e4b0be0a89bb0b07  511111504412           CAMPBELLS HOME STYLE   
9     5c408e8bcd244a1fdb47aee7  511111504788                           TEST   
10    5f4bf556be37ce0b4491554d  511111516354  TEST BRANDCODE @1598813526777   
11    57c08106e4b0718ff5fcb02c  511111102540                            NaN   
12    588ba07be4b02187f85cdadd  511111201076                        CALUMET   
13    5d6413156d5f3b23d1bc790a  511111205012                   511111205012   
14    585a9611e4b03e62d1ce0e74  511111801801              AUNT JEMIMA SYRUP   
15    57e5820ce4b0ac389136a311  511111202233                         MOLSON   
16    5fb6adb8be37ce522e165cb8  511111817376                       LOTRIMIN   
17    5f358338be37ce443bf9d55a  511111515319  TEST BRANDCODE @1597342520277   
18    5fb28549be37ce522e165cb5  511111317364                            NaN   
19    592486bfe410d61fcea3d139  511111300700                        ST IVES   
20    5c4699f387ff3577e203ea29  511111305125                     CHRISIMAGE   
21    5da6071ea60b87376833e34d  511111005650                   ALKA SELTZER   
22    57ebc011e4b0ac389136a335  511111802129         JACK DANIEL'S BARBECUE   
23    5332f5fee4b03c9a25efd0bd  511111303947                            NaN   
24    5332fa7ce4b03c9a25efd22e  511111802914                            NaN   
25    5e9f18bfbe37ce3e45b6a77f  511111914549                            NaN   
26    592486bee410d61fcea3d133  511111400769               MAGNUM Ice Cream   
27    5d66d71fa3a018093ab34728  511111105329                   511111105329   
28    5f493e72be37ce64d0ae36c6  511111316114  TEST BRANDCODE @1598635634882   
29    5f4936ddbe37ce52f8314fd9  511111315957                            NaN   
...                        ...           ...                            ...   
1137  57ebbfa9e4b0ac389136a331  511111202165                                  
1138  5a5d2c59e4b06ba572cf24a9  511111900054                        SUNSILK   
1139  5887a051e4b02187f85cdad0  511111401209                        BAKER'S   
1140  5ab1528ce4b0915382df761c  511111504344                     ECCE PANIS   
1141  5d66d6016d5f3b6188d4f047  511111405306                   511111405306   
1142  5f21c54abe37ce7958c59528  511111315230                       AIR WICK   
1143  5332f7fde4b03c9a25efd168  511111903345                         HAMM'S   
1144  57c08242e4b0718ff5fcb032  511111202516                            NaN   
1145  5f5bc4f1be37ce17125ac0e9  511111716594  TEST BRANDCODE @1599849713740   
1146  5332fa12e4b03c9a25efd1e6  511111703105                            NaN   
1147  5dc0351a1dda2c0ad7da64aa  511111306238                MONSTERS CEREAL   
1148  5d66e2d46d5f3b6188d4f056  511111905523              TAZO BOTTLED TEAS   
1149  5a0219ede4b00d6a5ccb394e  511111400554                       TONI&GUY   
1150  5d66da306d5f3b6188d4f04c  511111005377                   511111005377   
1151  5d66dfe6a3a018093ab3472c  511111905479                   511111905479   
1152  5f3e9172be37ce5a0e01b955  511111715559  TEST BRANDCODE @1597935986434   
1153  58b598dbe4b0857c2ddb7253  511111101017                       MICKEY'S   
1154  57d95826e4b0ac389136a2b7  511111302346                  CRYSTAL LIGHT   
1155  5a5d1bffe4b0db471c2d0423  511111000396               COORS EXTRA GOLD  

## creating a postgres database and inserting tables 

In [88]:
#Connecting my jupyter notebook to postgres database. I have created a new database name as 'fetch_rewards' in postgres
#(I have attached screenshot of my cmd window)

conn = psycopg2.connect(
    host="localhost",
    database="fetch_rewards",
    user="postgres",
    password="anuj13")

In [89]:
conn.autocommit = True
cursor = conn.cursor()

In [90]:
cursor.execute("SELECT VERSION();")
version = cursor.fetchone()
print("Connected to - ", version)

Connected to -  ('PostgreSQL 13.3, compiled by Visual C++ build 1914, 64-bit',)


In [91]:
#Created an engine using sqlalchemy in order to load data into my databse in postgres

engine  = sqlalchemy.create_engine('postgresql://postgres:anuj13@localhost:5432/fetch_rewards')

In [92]:
#Loading my 4 tables into database. 

receipts.to_sql('receipts', engine)
rewards_receipt.to_sql('rewards_receipt',engine)
users.to_sql('users', engine)
brands.to_sql('brands', engine)

ValueError: Table 'receipts' already exists.

## Till here I have cleaned all 4 dataset, created a database in postgres and loaded all the clean data to the "fetch_rewards" database. Now we have a clean database from where we can write sql query and get answers to the business questions

## Q) When considering average spend from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?

In [93]:
count = '''(select rewardsreceiptstatus, avg(totalSpent) from receipts group by rewardsreceiptstatus )'''

In [94]:
cursor.execute(count)
new = cursor.fetchall()
for n in new:
    print(n)

('REJECTED', 19.544970059880253)
('SUBMITTED', None)
('FLAGGED', 2635.5702469136068)
('PENDING', 28.03244897959184)
('FINISHED', 1244.372934121628)


## A) Looking at the data in above cell, the average spend from receipts with 'rewardsReceiptStatus’ = FINISED is 1244.373 and with 'rewardsReceiptStatus’ = REJECTED is 19.545, therefore for average spend from receipts with 'rewardsReceiptStatus’ = FINISED  is greater

## Q) When considering total number of items purchased from receipts with 'rewardsReceiptStatus’ of ‘Accepted’ or ‘Rejected’, which is greater?

In [95]:
item_purchased = '''(select rewardsreceiptstatus, sum(purchaseditemcount) from receipts group by rewardsreceiptstatus )'''

In [96]:
cursor.execute(item_purchased)
new = cursor.fetchall()
for n in new:
    print(n)

('REJECTED', 740.0)
('SUBMITTED', None)
('FLAGGED', 294816.0)
('PENDING', None)
('FINISHED', 1364998.0)


## A) Looking at the data in above cell, total number of items purchased from receipts with 'rewardsReceiptStatus’ = FINISED is 1364998.0 and with 'rewardsReceiptStatus’ = REJECTED is 740.0, therefore total number of items purchased from receipts with 'rewardsReceiptStatus’ = FINISED  is greater

# Data Quality Issues

# Communicate with Stakeholders

#### Question:
Construct an email or slack message that is understandable to a product or business leader who isn’t familiar with your day to day work. This part of the exercise should show off how you communicate and reason about data with others. Commit your answers to the git repository along with the rest of your exercise.

What questions do you have about the data?


How did you discover the data quality issues?
What do you need to know to resolve the data quality issues?
What other information would you need to help you optimize the data assets you're trying to create?
What performance and scaling concerns do you anticipate in production and how do you plan to address them?